sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,317 @@
+#define PLIC_UART_LAST_IRQ_ID kDifPlicUartRxParityErr
+#define PLIC_SPI_DEVICE_FIRST_IRQ_ID kDifPlicSpiDeviceRxF
+#define PLIC_SPI_DEVICE_LAST_IRQ_ID kDifPlicSpiDeviceTxUnderflow
+#define PLIC_FLASH_CONTROL_FIRST_IRQ_ID kDifPlicFlashCtrlProgEmpty
+#define PLIC_FLASH_CONTROL_LAST_IRQ_ID kDifPlicFlashCtrlOpError
+#define PLIC_HMAC_FIRST_IRQ_ID kDifPlicHmacDone
+#define PLIC_HMAC_LAST_IRQ_ID kDifPlicHmacErr
+#define PLIC_ALERT_HANDLER_FIRST_IRQ kDifPlicAlertHandlerClassA
+#define PLIC_ALERT_HANDLER_LAST_IRQ kDifPlicAlertHandlerClassD
+#define PLIC_NMI_GEN_FIRST_IRQ kDifPlicNmiGenEsc0
+#define PLIC_NMI_GEN_LAST_IRQ kDifPlicNmiGenEsc3
+
+// Peripheral IRQ sources that interrupt the Ibex core
+#define PLIC_IBEX_SOURCE_FIRST_IRQ kDifPlicGpio0
+#define PLIC_IBEX_SOURCE_LAST_IRQ kDifPlicNmiGenEsc3
+
+// This define is used to calculate the IRQ index in IP, LE, IE registers.
+// These registers are 32bit wide, and in order to accommodate all the IRQs,
+// multiple of the same type registers are defined (IE00, IE01, ...). For
+// example, IRQ ID 33 corresponds to bit 1 in registers IP1, LE1, IE01. The
+// value 33 is chosen instead of 32 because IRQ IDs start from 1 rather than 0.
+#define PLIC_ID_TO_INDEX_MODULO 33u
+
+static bool plic_priority_validate(uint32_t priority) {
+  return priority <= RV_PLIC_MAX_PRIORITY;
+}
+
+/**
+ * Check the PLIC state for whether an IRQ has been claimed
+ */
+static bool plic_irq_claimed(dif_plic_t *plic) {
+  return (plic->claimed_irq.peripheral != kDifPlicPeripheralNone &&
+          plic->claimed_irq.source != kDifPlicNoIrq);
+}
+
+static void plic_irq_claimed_data_clear(dif_plic_t *plic) {
+  plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+  plic->claimed_irq.source = kDifPlicNoIrq;
+}
+
+/**
+ * Get IE, IP or LE register offset (IE00, IE01, ...) from IRQ source ID.
+ */
+static ptrdiff_t plic_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq / PLIC_ID_TO_INDEX_MODULO;
+  return register_index * sizeof(uint32_t);
+}
+
+/**
+ * Get PRIO register offset (PRIO0, PRIO1, ...) from IRQ source ID.
+ */
+static ptrdiff_t plic_priority_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq;
+  return register_index * sizeof(uint32_t);
+}
+
+/**
+ * Get IE, IP, LE register bit index from IRQ source ID. IRQ IDs start from 1,
+ * however IRQ bit index inside the registers start from 0. This means that in
+ * order to get IRQ bit index in the register, 1 has to be substracted from the
+ * obtained remainder.
+ */
+static uint8_t plic_bit_index_from_irq_id(dif_plic_irq_src_id_t irq) {
+  return ((uint32_t)irq % PLIC_ID_TO_INDEX_MODULO) - 1;
+}
+
+/**
+ * Get the target that will be interrupted by the IRQ source
+ */
+static bool plic_irq_target_from_src_id(dif_plic_irq_src_id_t irq,
+                                        dif_plic_target_t *target) {
+  if (irq >= PLIC_IBEX_SOURCE_FIRST_IRQ && irq <= PLIC_IBEX_SOURCE_LAST_IRQ) {
```

Comment:
After reading the spec, I was under the impression that multiple targets could enable the same interrupt. I also read your notes and I agree that this must be clarified.